contracts/managers/allocators/BinaryAllocator.sol

```diff
@@ -0,0 +1,500 @@
+/*
+    Copyright 2019 Set Labs Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+pragma solidity 0.5.7;
+pragma experimental "ABIEncoderV2";
+
+import { ERC20Detailed } from "openzeppelin-solidity/contracts/token/ERC20/ERC20Detailed.sol";
+import { Math } from "openzeppelin-solidity/contracts/math/Math.sol";
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import { CommonMath } from "set-protocol-contracts/contracts/lib/CommonMath.sol";
+import { ICore } from "set-protocol-contracts/contracts/core/interfaces/ICore.sol";
+import { ISetToken } from "set-protocol-contracts/contracts/core/interfaces/ISetToken.sol";
+import { SetTokenLibrary } from "set-protocol-contracts/contracts/core/lib/SetTokenLibrary.sol";
+
+import { AllocatorMathLibrary } from "../lib/AllocatorMathLibrary.sol";
+import { FlexibleTimingManagerLibrary } from "../lib/FlexibleTimingManagerLibrary.sol";
+import { IAllocator } from "./IAllocator.sol";
+import { IOracle } from "../../meta-oracles/interfaces/IOracle.sol";
+
+
+/**
+ * @title BinaryAllocator
+ * @author Set Protocol
+ *
+ * Implementing IAllocator the BinaryAllocator flips between two all or nothing
+ * allocations of the base asset depending what allocation the calling manager is seeking. In
+ * addition, if either collateral Set becomes 4x more valuable than the other the contract will
+ * create a new collateral Set and use that Set going forward.
+ */
+contract BinaryAllocator is
+    IAllocator
+{
+    using SafeMath for uint256;
+
+    /* ============ Events ============ */
+
+    event NewCollateralLogged(
+        bytes32 indexed _hashId,
+        address _collateralAddress
+    );
+
+    /* ============ Constants ============ */
+    uint256 constant MINIMUM_COLLATERAL_NATURAL_UNIT_DECIMALS = 6;
+
+    /* ============ State Variables ============ */
+    ICore public coreInstance;
+    address public setTokenFactoryAddress;
+
+    ERC20Detailed public baseAssetInstance;
```

Comment:
No need to make any changes, but a suggestion for the future could be to create a struct for further succinctness

```
Struct Asset {
   IOracle oracle;
   uint8 decimals;
   ERC20Detailed instance;
}
```